docs/payments/payment-gateway-api.md

```diff
@@ -153,11 +153,14 @@ $order->payment_complete();
 
 This ensures stock reductions are made, and the status is changed to the correct value.
 
-If payment fails, you should throw an error and return null:
+If payment fails, you should throw an error and return an array with the failure status:
 
 ```php
-wc_add_notice( \_\_('Payment error:', 'woothemes') . $error_message, 'error' );
-return;
+wc_add_notice( __('Payment error:', 'woothemes') . $error_message, 'error' );
+return array(
+    'result'   => 'failure',
+    'redirect' => '', // you can redirect to a custom page here
```

Comment:
> I'm not 100% sure which code paths to follow, but historically, and AFAIK also currently, the redirect only matters upon success. 

I didn't know that the redirect was only used in the success flow @RadoslavGeorgiev, thanks a lot for pointing me in the good direction 🙇 . I was mistakenly assuming it would be used when it was set. I fully agree it does not make sense to include `'redirect' => ''` it in the docs.

> I'd prefer to look into checking whether redirect is set before accessing it, ex. [here](https://github.com/woocommerce/woocommerce/blob/fe2dd0be2babe6a11f0cfb16f7e2a655624ffcd7/plugins/woocommerce/src/StoreApi/Legacy.php#L71). It should be a fairly simple change, would you mind trying it out, and checking if tests pass?

I'm taking a slightly different approach in this [commit](https://github.com/woocommerce/woocommerce/pull/53156/commits/165e173bd9dd36ca9c8e1bb1e97d6c8d5993f2b5)  to use the null coallescing operator to fallback to an empty string. Also I I'm checking in the frontend (legacy) if the redirect exists before accessing it [77297a1](https://github.com/woocommerce/woocommerce/pull/53156/commits/77297a143c6ef798a1dfd8f3946c2d60f9397241). This is already [happening](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/base/context/providers/cart-checkout/checkout-processor.ts#L231) in the block checkout.

Please have a look and let me know what you think.